sequencer/src/state.rs

```diff
@@ -45,63 +43,102 @@ impl Default for ValidatedState {
+            "Invalid Block Root Error: {}, {}",
+            block_merkle_tree_root,
+            proposal.block_merkle_tree_root
+        )
+    );
 
-        let fee_merkle_tree_root = fee_merkle_tree.commitment();
-        anyhow::ensure!(
-            proposal.fee_merkle_tree_root == fee_merkle_tree_root,
-            anyhow::anyhow!(
-                "Invalid Fee Root Error: {}, {}",
-                fee_merkle_tree_root,
-                proposal.fee_merkle_tree_root
-            )
-        );
+    let fee_merkle_tree_root = fee_merkle_tree.commitment();
+    anyhow::ensure!(
+        proposal.fee_merkle_tree_root == fee_merkle_tree_root,
+        anyhow::anyhow!(
+            "Invalid Fee Root Error: {}, {}",
+            fee_merkle_tree_root,
+            proposal.fee_merkle_tree_root
+        )
+    );
+    Ok(())
+}
 
-        Ok(ValidatedState {
-            block_merkle_tree,
-            fee_merkle_tree,
-        })
+/// Fetch receipts from the l1 and add them to local balance.
+fn update_balance(fee_merkle_tree: &mut FeeMerkleTree, parent: &Header) {
+    let receipts = fetch_fee_receipts(parent);
+    for FeeReceipt { recipient, amount } in receipts {
+        // Get the balance in order to add amount, ignoring the proof.
+        match fee_merkle_tree.universal_lookup(recipient) {
+            LookupResult::Ok(balance, _) => fee_merkle_tree
+                .update(recipient, balance.add(amount))
+                .unwrap(),
+            // Handle `NotFound` and `NotInMemory` by initializing
+            // state.
+            _ => fee_merkle_tree.update(recipient, amount).unwrap(),
+        };
     }
 }
+/// Validate builder account by verifiying signature and charging the account.
+fn validate_builder(
+    fee_merkle_tree: &mut FeeMerkleTree,
+    proposed_header: &Header,
+) -> anyhow::Result<()> {
+    let mut verifiable_header = proposed_header.clone();
+    // These unraps should be safe since Header::new() must have
+    // set these fields.
+    let builder_signature = verifiable_header.builder_signature.take().unwrap();
+    let builder_address = verifiable_header.builder_address.take().unwrap();
+    let builder_fee_amount = verifiable_header.builder_fee_amount.take().unwrap();
+    let header_bytes = serde_json::to_string(&verifiable_header)
+        .unwrap()
+        .into_bytes();
+
+    // verify signature
+    anyhow::ensure!(
+        builder_signature
+            .verify(header_bytes, builder_address.address())
+            .is_ok(),
+        "Invalid Builder Signature"
+    );
+
+    // charge the fee to the builder
+    let mut fee_merkle_tree = fee_merkle_tree.clone();
+    match fee_merkle_tree.universal_lookup(builder_address) {
+        LookupResult::Ok(balance, _) => fee_merkle_tree
+            .update(builder_address, balance.sub(builder_fee_amount))
+            .unwrap(),
+        // `NotFound` or `NotInMemory` is a ghost account that
+        // somehow signed the block so it must be a BUG.
```

Comment:
Not true, any malicious builder can sign with an account that doesn't exist in an attempt to avoid paying a fee. For `NotFound`, returning an error is the correct thing to do (it's the same as not having enough balance). For `NotInMemory`, we will have to look up the account as part of catchup

Reply:
fixed in https://github.com/EspressoSystems/espresso-sequencer/commit/95fe801091f35ff6874efe52d24bd8215121872c